From 8579ebb26b6e0795f8b834b5713c7c3401e3fab2 Mon Sep 17 00:00:00 2001
From: Evan Wang <xswang@marvell.com>
Date: Tue, 23 Aug 2016 21:26:21 +0800
Subject: [PATCH 0880/2241] mvneta: add restart auto-negotiation supported

-This patch added support to restart port auto-neg,
 which is used by ethtool '-r' option.

Change-Id: I3b7df31476c4d61632fd21921b1e43b18df03566
Signed-off-by: Evan Wang <xswang@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/32132
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Wilson Ding <dingwei@marvell.com>
---
 drivers/net/ethernet/marvell/mvneta.c | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/drivers/net/ethernet/marvell/mvneta.c b/drivers/net/ethernet/marvell/mvneta.c
index 6d8185d..1aa3e64 100644
--- a/drivers/net/ethernet/marvell/mvneta.c
+++ b/drivers/net/ethernet/marvell/mvneta.c
@@ -3776,6 +3776,22 @@ static void mvneta_ethtool_get_regs(struct net_device *dev,
 	regs_buff[reg_index] = mvreg_read(pp, MVNETA_GMAC_AUTONEG_CONFIG);
 }
 
+static int mvneta_ethtool_nway_reset(struct net_device *dev)
+{
+	struct mvneta_port *pp = netdev_priv(dev);
+
+	if (!netif_running(dev))
+		return -EAGAIN;
+
+	if (!pp->phy_dev)
+		return -EOPNOTSUPP;
+
+	if (pp->phy_dev->autoneg == AUTONEG_DISABLE)
+		return -EINVAL;
+
+	return phy_start_aneg(pp->phy_dev);
+}
+
 static const struct net_device_ops mvneta_netdev_ops = {
 	.ndo_open            = mvneta_open,
 	.ndo_stop            = mvneta_stop,
@@ -3806,6 +3822,7 @@ const struct ethtool_ops mvneta_eth_tool_ops = {
 	.set_rxfh	= mvneta_ethtool_set_rxfh,
 	.get_regs_len	= mvneta_ethtool_get_regs_len,
 	.get_regs	= mvneta_ethtool_get_regs,
+	.nway_reset	= mvneta_ethtool_nway_reset,
 };
 
 /* Initialize hw */
-- 
2.7.4

